engine/access/rpc/backend/backend_accounts.go

```diff
@@ -62,43 +69,57 @@ func (b *backendAccounts) GetAccountAtBlockHeight(
 		return nil, rpc.ConvertStorageError(err)
 	}
 
-	// get block ID of the header at the given height
+	// get the block id of the latest sealed header
 	blockID := header.ID()
 
-	account, err := b.getAccountAtBlockID(ctx, address, blockID)
+	account, err := b.getAccountAtBlock(ctx, address, blockID, header.Height)
 	if err != nil {
+		b.log.Debug().Err(err).Msgf("failed to get account at blockID: %v", blockID)
 		return nil, err
 	}
 
 	return account, nil
 }
 
-func (b *backendAccounts) getAccountAtBlockID(
+// getAccountAtBlock returns the account details at the given block
+//
+// The data may be sourced from the local storage or from an execution node depending on the nodes's
+// configuration and the availability of the data.
+func (b *backendAccounts) getAccountAtBlock(
 	ctx context.Context,
 	address flow.Address,
 	blockID flow.Identifier,
+	height uint64,
 ) (*flow.Account, error) {
-
-	exeReq := &execproto.GetAccountAtBlockIDRequest{
-		Address: address.Bytes(),
-		BlockId: blockID[:],
+	if b.scriptExecMode == ScriptExecutionModeExecutionNodesOnly {
+		return b.getAccountFromAnyExeNode(ctx, address, blockID)
 	}
 
-	execNodes, err := executionNodesForBlockID(ctx, blockID, b.executionReceipts, b.state, b.log)
-	if err != nil {
-		return nil, rpc.ConvertError(err, "failed to get account from the execution node", codes.Internal)
-	}
+	account, err := b.getAccountFromLocalStorage(ctx, address, height)
 
-	var exeRes *execproto.GetAccountAtBlockIDResponse
-	exeRes, err = b.getAccountFromAnyExeNode(ctx, execNodes, exeReq)
-	if err != nil {
-		b.log.Error().Err(err).Msg("failed to get account from execution nodes")
-		return nil, rpc.ConvertError(err, "failed to get account from the execution node", codes.Internal)
+	if err != nil && b.scriptExecMode == ScriptExecutionModeFailover {
+		return b.getAccountFromAnyExeNode(ctx, address, blockID)
 	}
 
-	account, err := convert.MessageToAccount(exeRes.GetAccount())
+	return account, err
+}
+
+// getAccountFromLocalStorage retrieves the given account from the local storage.
+func (b *backendAccounts) getAccountFromLocalStorage(
+	ctx context.Context,
+	address flow.Address,
+	height uint64,
+) (*flow.Account, error) {
+	// make sure data is available for the requested block
+	account, err := b.scriptExecutor.GetAccountAtBlockHeight(ctx, address, height)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to convert account message: %v", err)
+		if errors.Is(err, ErrDataNotAvailable) {
```

Comment:
not directly since it has some additional error type handling done in `convertScriptExecutionError` that doesn't apply to getting accounts, but I could extract the common bits out to a separate function